### PR TITLE
nixos/sway: add programs.sway.extraOptions

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Needed after Optimus PR #66601
In fact sway runs on Intel card in such setup, but it detects nvidia drivers alongside and so refuses to start without magic argument

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @primeos

### CODE REVIEW

**Review by primeos — COMMENTED** [review]
Hm, tbh I'm not sure what do think of this. I guess this is only/mainly relevant when using the session file with a login manager.

It might be better to add an option like `extraOptions` (string or e.g. list of strings) to keep it more generic and add examples for `verbose`, `debug`, `unsupported-gpu` (proprietary AMD), and `my-next-gpu-wont-be-nvidia` ([Nvidia](https://itsfoss.com/wp-content/uploads/2012/09/Linus-Torvalds-Fuck-You-Nvidia.jpg)).

> In fact sway runs on Intel card in such setup, but it detects nvidia drivers alongside and so refuses to start without magic argument

If it does indeed run without proprietary drivers this could qualify for a bug report I guess (maybe one should ask very politely though, especially if the name of a certain company (it who must not be named) is mentioned, which might cause "discomfort" for some readers).

**Comment by gnidorah** [comment]
@primeos 

Thank you much for the answer!

> I guess this is only/mainly relevant when using the session file with a login manager.

Mainly for that, right.

> It might be better to add an option like extraOptions (string or e.g. list of strings) to keep it more generic and add examples for verbose, debug, unsupported-gpu (proprietary AMD), and my-next-gpu-wont-be-nvidia (Nvidia).

Will do! Thanks for suggestion.

> If it does indeed run without proprietary drivers this could qualify for a bug report I guess (maybe one should ask very politely though, especially if the name of a certain company (it who must not be named) is mentioned, which might cause "discomfort" for some readers).

Yes, it runs on Intel card. I've created an upstream issue, but I doubt they will invest time into this https://github.com/swaywm/sway/issues/4838

File: pkgs/applications/window-managers/sway/wrapper.nix [hunk 1]
```diff
@@ -41,7 +42,7 @@ in symlinkJoin {
     wrapProgram $out/bin/sway \
       --prefix PATH : "${swaybg}/bin" ${optionalString withGtkWrapper ''\
         "''${gappsWrapperArgs[@]}"
-      ''}
+      ''} ${optionalString withNvidia "--add-flags --my-next-gpu-wont-be-nvidia"}
```
[the remark this answers is not in the document]
Comment (gnidorah) [reply pkgs/applications/window-managers/sway/wrapper.nix]: @primeos Yes it somehow worked :smile:

**Comment by lovesegfault** [comment]
This is what I used to do to support Bumblebee:
https://github.com/lovesegfault/dotfiles/blob/1a4049e2454543e6f3b856b64432fa63a38d447e/bin/wm#L33-L42

It worked fine at finding an appropriate GPU and forcing it as the Sway one.

**Comment by Elyhaka** [comment]
I agree with @primeos: we should not support wrapper options explicitly that are not recommended and/or supported by upstream. On the other hand, I think that providing a `extraOptions` is a better approach and will give the same result in the end.

**Comment by gnidorah** [comment]
@primeos ping

**Review by primeos — APPROVED** [review]
@gnidorah sorry for the delay and thanks for the ping :)

I gave it a test and the build did fail if the GTK wrapper was enabled:
```
/build/.attr-0: line 11: --add-flags: command not found
```

I pushed a fix that tries to simplify the `wrapProgram` mess I've created... :o :D
(And I've added an extra note about opening issues with proprietary drivers as upstream would most likely want us to do this.)
Are you ok with these changes?

**Comment by gnidorah** [comment]
> Are you ok with these changes?

Sure. Thank you!

**Comment by primeos** [comment]
Merged, thanks for the PR :)

**Comment by gnidorah** [comment]
@primeos Thank you!

**Pull request merged**